kong/router/atc.lua

```diff
@@ -318,25 +317,44 @@ end
 -- split port in host, ignore form '[...]'
 -- example.com:123 => example.com, 123
 -- example.*:123 => example.*, 123
-local function split_host_port(h)
-  if not h then
-    return nil, nil
-  end
+local split_host_port
+do
+  local tonumber = tonumber
 
-  local p = h:find(":", nil, true)
-  if not p then
-    return h, nil
-  end
+  local memo_h = setmetatable({}, { __mode = "k" })
+  local memo_p = setmetatable({}, { __mode = "k" })
```

Comment:
I know this code come from the old router, but it seems to have some issues:

By making the keys weak, the table's value are dependent on temporary objects created during requests that are gone by the end of the request. This means that these table are probably not caching enough content as they should.

Further, the logic below seems to assume that if `memo_h[h]` is not `nil`, then `memo_p[h]` must also contains a valid value. This may not be true if Lua GC'ed `memo_p[h]` before GCing `memo_h[h]`.

I think we should change this (and the traditional router) to use proper LRU Cache instead of this pattern. Otherwise in rare cases it may produce very hard to debug issues.

Reply:
It is a reasonable concern. I will change weak table to lru cache.